File: nars_java/nars/language/IntersectionExt.java
/*
 * IntersectionExt.java
 *
 * Copyright (C) 2008  Pei Wang
 *
 * This file is part of Open-NARS.
 *
 * Open-NARS is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Open-NARS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Open-NARS.  If not, see <http://www.gnu.org/licenses/>.
 */
package nars.language;

import java.util.Arrays;
import java.util.Set;
import java.util.TreeSet;
import nars.core.Parameters;
import nars.io.Symbols.NativeOperator;
import static nars.language.SetTensional.verifySortedAndUnique;

/**
 * A compound term whose extension is the intersection of the extensions of its term
 */
public class IntersectionExt extends CompoundTerm {

    /**
     * Constructor with partial values, called by make
     * @param n The name of the term
     * @param arg The component list of the term
     */
    private IntersectionExt(Term[] arg) {
        super(arg);
        
        if (Parameters.DEBUG) {
            verifySortedAndUnique(arg, false);
        }
        
    }


    /**
     * Clone an object
     * @return A new object, to be casted into a IntersectionExt
     */
    @Override
    public IntersectionExt clone() {
        return new IntersectionExt(term);
    }
    
    @Override
    public CompoundTerm clone(Term[] replaced) {
        if (replaced.length == 1)
            return (CompoundTerm) replaced[0];
        else if (replaced.length > 1)
            return (CompoundTerm) make(replaced);
        else
            throw new RuntimeException("Invalid # of terms for Intersection: " + Arrays.toString(replaced));
    }
    
    /**
     * Try to make a new compound from two term. Called by the inference rules.
     * @param term1 The first compoment
     * @param term2 The first compoment
     * @param memory Reference to the memory
     * @return A compound generated or a term it reduced to
     */
    public static Term make(Term term1, Term term2) {
        Set<Term> set;
        if ((term1 instanceof SetInt) && (term2 instanceof SetInt)) {
            set = new TreeSet<>(((CompoundTerm) term1).getTermList());
            set.addAll(((CompoundTerm) term2).getTermList());        // set union
            return SetInt.make(set);
        }
        if ((term1 instanceof SetExt) && (term2 instanceof SetExt)) {
            set = new TreeSet<>(((CompoundTerm) term1).getTermList());
            set.retainAll(((CompoundTerm) term2).getTermList());     // set intersection
            return SetExt.make(set);
        }
        if (term1 instanceof IntersectionExt) {
            set = new TreeSet<>(((CompoundTerm) term1).getTermList());
            if (term2 instanceof IntersectionExt) {
                // (&,(&,P,Q),(&,R,S)) = (&,P,Q,R,S)
                set.addAll(((CompoundTerm) term2).getTermList());
            }               
            else {
                // (&,(&,P,Q),R) = (&,P,Q,R)
                set.add(term2);
            }               
        } else if (term2 instanceof IntersectionExt) {
            // (&,R,(&,P,Q)) = (&,P,Q,R)
            set = new TreeSet<>(((CompoundTerm) term2).getTermList());
            set.add(term1);
        } else {
            set = new TreeSet<>();
            set.add(term1);
            set.add(term2);
        }
        return make(set.toArray(new Term[set.size()]));
    }

    public static Term make(Term... t) {
        if (t.length == 1) return t[0]; // special case: single component                
        return new IntersectionExt(Term.toSortedSetArray(t));
    }



    /**
     * Get the operator of the term.
     * @return the operator of the term
     */
    @Override
    public NativeOperator operator() {
        return NativeOperator.INTERSECTION_EXT;
    }

    /**
     * Check if the compound is communitative.
     * @return true for communitative
     */
    @Override
    public boolean isCommutative() {
        return true;
    }
}

File: nars_scala/src/main/scala/nars/language/IntersectionExt.scala
package nars.logic.language

import java.util._
import nars.io.Symbols
import nars.storage.Memory
import IntersectionExt._
//remove if not needed
import scala.collection.JavaConversions._
import CompoundTerm._

object IntersectionExt {

  /**
   * Try to make a new compound from two components. Called by the inference rules.
   * @param term1 The first compoment
   * @param term2 The first compoment
   * @param memory Reference to the memory
   * @return A compound generated or a term it reduced to
   */
  def make(term1: Term, term2: Term, memory: Memory): Term = {
    var set: TreeSet[Term] = null
    if ((term1.isInstanceOf[SetInt]) && (term2.isInstanceOf[SetInt])) {
      set = new TreeSet[Term](term1.asInstanceOf[CompoundTerm].cloneComponents())
      set.addAll(term2.asInstanceOf[CompoundTerm].cloneComponents())
      return SetInt.make(set, memory)
    }
    if ((term1.isInstanceOf[SetExt]) && (term2.isInstanceOf[SetExt])) {
      set = new TreeSet[Term](term1.asInstanceOf[CompoundTerm].cloneComponents())
      set.retainAll(term2.asInstanceOf[CompoundTerm].cloneComponents())
      return SetExt.make(set, memory)
    }
    if (term1.isInstanceOf[IntersectionExt]) {
      set = new TreeSet[Term](term1.asInstanceOf[CompoundTerm].cloneComponents())
      if (term2.isInstanceOf[IntersectionExt]) {
        set.addAll(term2.asInstanceOf[CompoundTerm].cloneComponents())
      } else {
        set.add(term2.clone().asInstanceOf[Term])
      }
    } else if (term2.isInstanceOf[IntersectionExt]) {
      set = new TreeSet[Term](term2.asInstanceOf[CompoundTerm].cloneComponents())
      set.add(term1.clone().asInstanceOf[Term])
    } else {
      set = new TreeSet[Term]()
      set.add(term1.clone().asInstanceOf[Term])
      set.add(term2.clone().asInstanceOf[Term])
    }
    make(set, memory)
  }

  /**
   * Try to make a new IntersectionExt. Called by StringParser.
   * @return the Term generated from the arguments
   * @param argList The list of components
   * @param memory Reference to the memory
   */
  def make(argList: ArrayList[Term], memory: Memory): Term = {
    val set = new TreeSet[Term](argList)
    make(set, memory)
  }

  /**
   * Try to make a new compound from a set of components. Called by the public make methods.
   * @param set a set of Term as compoments
   * @param memory Reference to the memory
   * @return the Term generated from the arguments
   */
  def make(set: TreeSet[Term], memory: Memory): Term = {
    if (set.size == 1) {
      return set.first()
    }
    val argument = new ArrayList[Term](set)
    val name = makeCompoundName(Symbols.INTERSECTION_EXT_OPERATOR, argument)
    val t = memory.nameToListedTerm(name)
    if ((t != null)) t else new IntersectionExt(argument)
  }
}

/**
 * A compound term whose extension is the intersection of the extensions of its components
 */
class IntersectionExt private (arg: ArrayList[Term]) extends CompoundTerm(arg) {

  /**
   * Constructor with full values, called by clone
   * @param n The name of the term
   * @param cs Component list
   * @param open Open variable list
   * @param i Syntactic complexity of the compound
   */
  private def this(n: String, 
      cs: ArrayList[Term], 
      con: Boolean, 
      i: Short) {
//    super(n, cs, con, i)
    this(cs)
    setName(n)
    this.isConstant_ = con
    this.complexity = i
  }

  /**
   * Clone an object
   * @return A new object, to be casted into a IntersectionExt
   */
  override def clone(): AnyRef = {
    new IntersectionExt(name, cloneList(components).asInstanceOf[ArrayList[Term]], isConstant_, complexity)
  }

  /**
   * Get the operator of the term.
   * @return the operator of the term
   */
  def operator(): String = Symbols.INTERSECTION_EXT_OPERATOR

  /**
   * Check if the compound is communitative.
   * @return true for communitative
   */
  override def isCommutative(): Boolean = true
}
